File: pkg/splprime/scan/tokentype.go
//go:generate stringer -type=TokenType
package scan

import "errors"

type TokenType int

const (
	Unknown TokenType = iota

	/* OPERATORS */
	OpPlus
	OpMinus
	OpStar
	OpSlash
	OpAssign
	OpEqual
	OpNotEqual
	OpGreaterThan
	OpLessThan
	OpGeq // Greater than or equal
	OpLeq // Less than or equal
	OpBang

	/* SPECIAL CHARS */
	CharSemicolon
	CharLParen
	CharRParen
	CharLBrace
	CharRBrace

	/* KEYWWORDS */
	KeywordTrue
	KeywordFalse
	KeywordAnd
	KeywordOr
	KeywordVar
	KeywordPrint
	KeywordIf
	KeywordElse
	KeywordWhile

	/* LITERALS */
	LitString
	LitNumber

	/* IDENTIFIERS */
	Ident

	/* EOF */
	EOF
)

func ResolveKeywordOrUnkown(s string) TokenType {
	switch s {
	case "true":
		return KeywordTrue
	case "false":
		return KeywordFalse
	case "and":
		return KeywordAnd
	case "or":
		return KeywordOr
	case "var":
		return KeywordVar
	case "print":
		return KeywordPrint
	case "if":
		return KeywordIf
	case "else":
		return KeywordElse
	case "while":
		return KeywordWhile
	default:
		return Unknown
	}
}

func ResolveKeywordOrIdent(s string) TokenType {
	t := ResolveKeywordOrUnkown(s)

	if t == Unknown {
		return Ident
	}

	return t
}

func ResolveStaticToken(s string) (TokenType, error) {
	switch s {
	// OPERATORS
	case "+":
		return OpPlus, nil
	case "-":
		return OpMinus, nil
	case "*":
		return OpStar, nil
	case "/":
		return OpSlash, nil
	case "=":
		return OpAssign, nil
	case "==":
		return OpEqual, nil
	case "!=":
		return OpNotEqual, nil
	case ">":
		return OpGreaterThan, nil
	case ">=":
		return OpGeq, nil
	case "<":
		return OpLessThan, nil
	case "<=":
		return OpLeq, nil
	case "!":
		return OpBang, nil

	// SPECIAL CHARS
	case ";":
		return CharSemicolon, nil
	case "(":
		return CharLParen, nil
	case ")":
		return CharRParen, nil
	case "{":
		return CharLBrace, nil
	case "}":
		return CharRBrace, nil

	default:
		t := ResolveKeywordOrUnkown(s)
		if t == Unknown {
			return Unknown, errors.New("invalid static token: " + s)
		} else {
			return t, nil
		}
	}
}

File: src/splprime/scan/TokenType.java
package splprime.scan;

public enum TokenType {
	/* OPERATORS */
	OP_PLUS,
	OP_MINUS,
	OP_STAR,
	OP_SLASH,
	OP_ASSIGN,
	OP_EQUAL,
	OP_NOT_EQUAL,
	OP_GREATER_THAN,
	OP_LESS_THAN,
	OP_GEQ, // Greater than or equal
	OP_LEQ, // Less than or equal
	OP_BANG,

	/* SPECIAL CHARS */
	CHAR_SEMICOLON,
	CHAR_L_PAREN,
	CHAR_R_PAREN,
	CHAR_L_BRACE,
	CHAR_R_BRACE,

	/* KEYWWORDS */
	KEYWORD_TRUE,
	KEYWORD_FALSE,
	KEYWORD_AND,
	KEYWORD_OR,
	KEYWORD_VAR,
	KEYWORD_PRINT,
	KEYWORD_IF,
	KEYWORD_ELSE,
	KEYWORD_WHILE,

	/* LITERALS */
	LIT_STRING,
	LIT_NUMBER,

	/* IDENTIFIERS */
	IDENT,

	/* EOF */
	EOF,
}

class TokenResolver {
	static TokenType resolveKeywordOrNull(String s) {
		switch (s) {
			case "true":
				return TokenType.KEYWORD_TRUE;
			case "false":
				return TokenType.KEYWORD_FALSE;
			case "and":
				return TokenType.KEYWORD_AND;
			case "or":
				return TokenType.KEYWORD_OR;
			case "var":
				return TokenType.KEYWORD_VAR;
			case "print":
				return TokenType.KEYWORD_PRINT;
			case "if":
				return TokenType.KEYWORD_IF;
			case "else":
				return TokenType.KEYWORD_ELSE;
			case "while":
				return TokenType.KEYWORD_WHILE;
			default:
				return null;
		}
	}

	static TokenType resolveKeywordOrIdent(String s) {
		var t = resolveKeywordOrNull(s);
		if (t == null) {
			return TokenType.IDENT;
		}

		return t;
	}

	static TokenType resolveStatic(String s) {

		switch (s) {
			/* OPERATORS */
			case "+":
				return TokenType.OP_PLUS;
			case "-":
				return TokenType.OP_MINUS;
			case "*":
				return TokenType.OP_STAR;
			case "/":
				return TokenType.OP_SLASH;
			case "=":
				return TokenType.OP_ASSIGN;
			case "==":
				return TokenType.OP_EQUAL;
			case "!=":
				return TokenType.OP_NOT_EQUAL;
			case ">":
				return TokenType.OP_GREATER_THAN;
			case ">=":
				return TokenType.OP_GEQ;
			case "<":
				return TokenType.OP_LESS_THAN;
			case "<=":
				return TokenType.OP_LEQ;
			case "!":
				return TokenType.OP_BANG;

			/* SPECIAL CHARS */
			case ";":
				return TokenType.CHAR_SEMICOLON;
			case "(":
				return TokenType.CHAR_L_PAREN;
			case ")":
				return TokenType.CHAR_R_PAREN;
			case "{":
				return TokenType.CHAR_L_BRACE;
			case "}":
				return TokenType.CHAR_R_BRACE;

			default:
				var t = resolveKeywordOrNull(s);
				if (t == null)
					throw new RuntimeException("invalid static token: " + s);
				else
					return t;
		}
	}
}
